### PR TITLE
CNX-9220 Revit Room to Speckle support

### PR DESCRIPTION
**Pull request by AlanRynne**
https://spockle.atlassian.net/browse/CNX-9220

Added Revit Room converter and auxiliary BoundarySegments conversion to PolyCurve.

⚠️ This PR depends on the Curve converter being complete. It currently lacks support for other basic curve types and only simple "box rooms" will be sent.

### CODE REVIEW

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  public BoundarySegmentConversionToSpeckle(IRawConversion<DB.Curve, ICurve> curveConverter)
+  {
+    _curveConverter = curveConverter;
+  }
+
+  public SOG.Polycurve RawConvert(IList<DB.BoundarySegment> target)
+  {
+    if (!target.Any())
+    {
+      throw new ArgumentException("Input Boundary segment list must at least have 1 segment");
```
Comment (connorivy) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs:19]: Related to a comment on Ian's PR. Is this supposed to be a SpeckleConversionException?

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  private readonly IRawConversion<DB.Curve, ICurve> _curveConverter;
+
+  public BoundarySegmentConversionToSpeckle(IRawConversion<DB.Curve, ICurve> curveConverter)
+  {
+    _curveConverter = curveConverter;
+  }
+
+  public SOG.Polycurve RawConvert(IList<DB.BoundarySegment> target)
+  {
+    if (!target.Any())
```
Comment (connorivy) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs:17]: very small thing, but the compiler will suggest that you prefer Count == 0 to .Any() for lists or arrays

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    var poly = new SOG.Polycurve();
+    foreach (var segment in target)
+    {
+      DB.Curve revitCurve = segment.GetCurve();
+      var curve = _curveConverter.RawConvert(revitCurve);
+
+      // POC: Not sure why we would need these to have an ElementID attached.
+      // curve["elementId"] = segment.ElementId.ToString();
```
Comment (connorivy) [inline DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs:29]: I agree that this information does not seem useful

**Review by connorivy — APPROVED** [review]
some minor comments, but looks good to me

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs (hunk 1 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs:19]: I don't think it should... we didn't actually "failed to convert" but rather "the input provided was not valid" which is not really the same.

I'm not very particular doing one or the other, we need to decide on a consistent strategy for this.

File: DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs (hunk 3 above)
Comment (AlanRynne) [reply DUI3-DX/Converters/Revit/Speckle.Converters.RevitShared/Raw/BoundarySegmentConversionToSpeckle.cs:29]: Ok... I'm going to leave that commented out for now, but with a better comment :)

**Comment by AlanRynne** [comment]
Rooms coming through after the curve conversions were added! 🚀
Screenshot 2024-05-02 at 12 06 58

**Pull request merged**